File: blaise-graph-theory/src/main/java/com/googlecode/blaisemath/graph/generate/StarGraphGenerator.java
package com.googlecode.blaisemath.graph.generate;

/*
 * #%L
 * BlaiseGraphTheory
 * --
 * Copyright (C) 2009 - 2022 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.google.common.graph.Graph;
import com.googlecode.blaisemath.graph.GraphUtils;

import java.util.AbstractList;

/**
 * Constructs star graph with n nodes; all nodes are connected to a central hub.
 * @author Elisha Peterson
 */
public final class StarGraphGenerator extends AbstractGraphGenerator {

    public StarGraphGenerator() {
        super("Star Graph");
    }

    @Override
    public Graph<Integer> apply(DefaultGeneratorParameters parameters) {
        final int nodes = parameters.getNodeCount();
        if (nodes == 0) {
            return GraphUtils.emptyGraph(parameters.isDirected());
        }
        return GraphGenerators.createGraphWithEdges(parameters, new AbstractList<Integer[]>() {
            @Override
            public Integer[] get(int index) {
                return new Integer[]{0, index + 1};
            }

            @Override
            public int size() {
                return nodes - 1;
            }
        });
    }
    
}

File: blaise-graph-theory/src/main/java/com/googlecode/blaisemath/graph/generate/StarGraphGenerator.kt
package com.googlecode.blaisemath.graph.generate

import com.google.common.graph.Graph
import com.googlecode.blaisemath.graph.ContractedGraphTest
import com.googlecode.blaisemath.graph.GraphUtils
import com.googlecode.blaisemath.graph.GraphUtilsTest
import com.googlecode.blaisemath.graph.SubgraphTest
import com.googlecode.blaisemath.graph.generate.GraphGrowthRule
import com.googlecode.blaisemath.graph.generate.GraphSeedRule
import com.googlecode.blaisemath.graph.generate.HopGrowthRule
import com.googlecode.blaisemath.graph.layout.SpringLayoutPerformanceTest
import com.googlecode.blaisemath.graph.metrics.AdditiveSubsetMetricTest
import com.googlecode.blaisemath.graph.metrics.BetweenCentralityTest
import com.googlecode.blaisemath.graph.metrics.ClosenessCentralityTest
import com.googlecode.blaisemath.graph.metrics.CooperationMetric
import com.googlecode.blaisemath.graph.metrics.EigenCentralityTest
import com.googlecode.blaisemath.graph.metrics.GraphCentralityTest
import com.googlecode.blaisemath.graph.metrics.SubsetMetricsTest
import com.googlecode.blaisemath.test.AssertUtils
import org.junit.BeforeClass
import java.util.*

/*
* #%L
* BlaiseGraphTheory
* --
* Copyright (C) 2009 - 2021 Elisha Peterson
* --
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* #L%
*/ /**
 * Constructs star graph with n nodes; all nodes are connected to a central hub.
 * @author Elisha Peterson
 */
class StarGraphGenerator : AbstractGraphGenerator("Star Graph") {
    override fun apply(parameters: DefaultGeneratorParameters?): Graph<Int?>? {
        val nodes = parameters.getNodeCount()
        return if (nodes == 0) {
            GraphUtils.emptyGraph(parameters.isDirected())
        } else GraphGenerators.createGraphWithEdges(parameters, object : AbstractList<Array<Int?>?>() {
            override fun get(index: Int): Array<Int?>? {
                return arrayOf(0, index + 1)
            }

            override fun size(): Int {
                return nodes - 1
            }
        })
    }
}